sky/check.py

```diff
@@ -43,22 +46,43 @@ def check_one_cloud(cloud_tuple: Tuple[str, clouds.Cloud]) -> None:
             if reason is not None:
                 echo(f'    Hint: {reason}')
         else:
+            disabled_clouds.append(cloud_repr)
             echo(f'    Reason: {reason}')
 
-    clouds_to_check = [
-        (repr(cloud), cloud) for cloud in clouds.CLOUD_REGISTRY.values()
-    ]
-    clouds_to_check.append(('Cloudflare, for R2 object store', cloudflare))
+    if clouds is not None:
+        clouds_to_check: List[Tuple[str, Any]] = []
+        for cloud in clouds:
+            if cloud.lower() == 'cloudflare':
+                clouds_to_check.append(
+                    ('Cloudflare, for R2 object store', cloudflare))
+            else:
+                clouds_to_check.append(
+                    (cloud.lower(), sky_clouds.CLOUD_REGISTRY.from_str(cloud)))
+    else:
+        clouds_to_check = [(repr(cloud), cloud)
+                           for cloud in sky_clouds.CLOUD_REGISTRY.values()]
+        clouds_to_check.append(('Cloudflare, for R2 object store', cloudflare))
 
     for cloud_tuple in sorted(clouds_to_check):
         check_one_cloud(cloud_tuple)
 
-    # Cloudflare is not a real cloud in clouds.CLOUD_REGISTRY, and should not be
-    # inserted into the DB (otherwise `sky launch` and other code would error
-    # out when it's trying to look it up in the registry).
+    # Cloudflare is not a real cloud in sky_clouds.CLOUD_REGISTRY, and should
+    # not be inserted into the DB (otherwise `sky launch` and other code would
+    # error out when it's trying to look it up in the registry).
     enabled_clouds = [
         cloud for cloud in enabled_clouds if not cloud.startswith('Cloudflare')
     ]
+    disabled_clouds = [
+        cloud for cloud in disabled_clouds if not cloud.startswith('Cloudflare')
+    ]
+    previously_enabled_clouds = [
+        repr(cloud) for cloud in global_user_state.get_cached_enabled_clouds()
+    ]
+    for cloud in previously_enabled_clouds:
+        is_not_duplicate = cloud not in enabled_clouds
+        is_not_disabled = cloud not in disabled_clouds
+        if is_not_duplicate and is_not_disabled:
+            enabled_clouds.append(cloud)
```

Comment:
nit: is it possible to use some set operations instead of manually looping the lists?